util/stmtsummary/statement_summary.go

```diff
@@ -0,0 +1,213 @@
+	sampleSQL       string
+	execCount       uint64
+	sumLatency      uint64
+	maxLatency      uint64
+	minLatency      uint64
+	sumAffectedRows uint64
+	// Number of rows sent to client.
+	sumSentRows uint64
+	// The first time this type of SQL executes.
+	firstSeen time.Time
+	// The last time this type of SQL executes.
+	lastSeen time.Time
+}
+
+// StmtExecInfo records execution status of each statement.
+type StmtExecInfo struct {
+	SchemaName    string
+	OriginalSQL   string
+	NormalizedSQL string
+	Digest        string
+	TotalLatency  uint64
+	AffectedRows  uint64
+	// Number of rows sent to client.
+	SentRows  uint64
+	StartTime time.Time
+}
+
+// NewStmtSummaryByDigest creates a stmtSummaryByDigest.
+func NewStmtSummaryByDigest() *stmtSummaryByDigest {
+	maxStmtCount := config.GetGlobalConfig().StmtSummary.MaxStmtCount
+	return &stmtSummaryByDigest{
+		summaryMap: kvcache.NewSimpleLRUCache(maxStmtCount, 0, 0),
+	}
+}
+
+// Convert StmtExecInfo to stmtSummary
+func convertStmtExecInfoToSummary(sei *StmtExecInfo) *stmtSummary {
+	// Trim SQL to size MaxSQLLength
+	maxSQLLength := config.GetGlobalConfig().StmtSummary.MaxSQLLength
+	normalizedSQL := sei.NormalizedSQL
+	if len(normalizedSQL) > int(maxSQLLength) {
+		normalizedSQL = normalizedSQL[:maxSQLLength]
+	}
+	sampleSQL := sei.OriginalSQL
+	if len(sampleSQL) > int(maxSQLLength) {
+		sampleSQL = sampleSQL[:maxSQLLength]
+	}
+
+	return &stmtSummary{
+		schemaName:      sei.SchemaName,
+		digest:          sei.Digest,
+		normalizedSQL:   normalizedSQL,
+		sampleSQL:       sampleSQL,
+		execCount:       1,
+		sumLatency:      sei.TotalLatency,
+		maxLatency:      sei.TotalLatency,
+		minLatency:      sei.TotalLatency,
+		sumAffectedRows: sei.AffectedRows,
+		sumSentRows:     sei.SentRows,
+		firstSeen:       sei.StartTime,
+		lastSeen:        sei.StartTime,
+	}
+}
+
+// Add a new StmtExecInfo to stmtSummary
+func addStmtExecInfoToSummary(sei *StmtExecInfo, ss *stmtSummary) {
+	ss.Lock()
+
+	ss.sumLatency += sei.TotalLatency
+	ss.execCount++
+	if sei.TotalLatency > ss.maxLatency {
+		ss.maxLatency = sei.TotalLatency
+	}
+	if sei.TotalLatency < ss.minLatency {
+		ss.minLatency = sei.TotalLatency
+	}
+	// TODO: uint64 overflow
+	ss.sumAffectedRows += sei.AffectedRows
+	ss.sumSentRows += sei.SentRows
+	if sei.StartTime.Before(ss.firstSeen) {
+		ss.firstSeen = sei.StartTime
+	}
+	if ss.lastSeen.Before(sei.StartTime) {
+		ss.lastSeen = sei.StartTime
+	}
+
+	ss.Unlock()
+}
+
+// Add a statement to statement summary.
+func (ss *stmtSummaryByDigest) AddStatement(sei *StmtExecInfo) {
+	key := &stmtSummaryCacheKey{
+		schemaName: sei.SchemaName,
+		digest:     sei.Digest,
+	}
+
+	var summary *stmtSummary
+	ss.Lock()
+	value, ok := ss.summaryMap.Get(key)
+	if ok {
+		summary = value.(*stmtSummary)
+	} else {
+		newSummary := convertStmtExecInfoToSummary(sei)
+		ss.summaryMap.Put(key, newSummary)
+	}
+	ss.Unlock()
+
+	// Lock a single entry, not the whole cache.
+	if summary != nil {
```

Comment:
Seems `summary` can not be nil.

Reply:
If an entry is not found in the cache, we create `newSummary`, so `summary` stays nil.

Reply:
Then why not define `summary` and `addStmtExecInfoToSummary` in the `ok` block?

Reply:
Or we can create a new summary without adding the stats, call addStmtExecInfoToSummary later.

Reply:
Separate `addStmtExecInfoToSummary` and the `ok` block because I want to move `addStmtExecInfoToSummary` out of the big cache lock, and add a smaller lock to it.
Why do I need to create a new summary?

Reply:
So we can remove the `if summary != nil ` check

Reply:
The code is simplified a little.